lib/Serialization/Serialization.cpp

```diff
@@ -2315,6 +2315,15 @@ void Serializer::writeForeignErrorConvention(const ForeignErrorConvention &fec){
                                            resultTypeID);
 }
 
+static SmallVector<Type, 2> collectDependenciesFromType(CanType ty) {
+  SmallVector<Type, 2> result;
+  ty.visit([&](CanType next) {
+    if (auto *nominal = next->getAnyNominal())
+      result.push_back(nominal->getDeclaredType());
```

Comment:
How about `getDeclaredInterfaceType` here, so we don't have any archetypes in there that would force deserialization of a generic environment?